x-pack/plugins/security_solution/server/endpoint/services/actions/utils.ts

```diff
@@ -431,3 +433,26 @@ export const formatEndpointActionResults = (
       })
     : [];
 };
+
+export const getAgentsMetadataInfo = async ({
+  esClient,
+  searchedAgentIds,
+  metadataService,
+}: {
+  esClient: ElasticsearchClient;
+  searchedAgentIds: string[];
+  metadataService: EndpointMetadataService;
+}): Promise<{ [k: string]: string }> => {
+  // get host metadata docs with queried agents
+  const metaDataDocs = await metadataService.findHostMetadataForFleetAgents(esClient, [
+    ...new Set(searchedAgentIds),
+  ]);
+  // agent ids and names from metadata
+  // map this into an object as {id1: name1, id2: name2} etc
+  const agentsMetadataInfo = metaDataDocs.reduce<{ [k: string]: string }>((acc, curr) => {
```

Comment:
So the return value might not actually have all of the `id`'s of the agents in it, because some might no longer be around. You need to use `searchAgentIds` in the loop here (for the `.reduce()`) and grab the data from the `metaDataDocs` if it is available.

Reply:
https://github.com/elastic/kibana/pull/139379/commits/101408939af5c38e5e2045e8afbfe99b544f9f53#diff-1dc1c196c034860eea2b5b152f8235e5455241766a270553f9d6dd2cb8c32ce5L452-R453